nexus/types/src/deployment.rs

```diff
@@ -97,6 +101,47 @@ impl SledResources {
     }
 }
 
+/// Policy and database inputs to the Reconfigurator planner
+///
+/// The primary inputs to the planner are the parent (either a parent blueprint
+/// or an inventory collection) and this structure. This type holds the
+/// fleet-wide policy as well as any additional information fetched from CRDB
+/// that the planner needs to make decisions.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PlanningInput {
```

Comment:
tl;dr: Don't feel like you have to do this now, but I wonder if we should:

- move some of the `Policy` stuff into `PlanningInput` (the sled state, zpools, and subnet)
- move the internal and external DNS generations that are accepted by all the same functions into this struct
- create some accessors/iterators around PlanningInput (e.g., methods to iterate the sleds) so that we don't have to make tons of tiny changes when we move stuff like this around inside this struct
- create a builder interface so that callers don't have to assemble all this by hand

I think this might be much more comprehensible and also eliminate quite a lot of code today that assembles these structs and pokes through their innards.  I'm thinking of the various bits in this PR where you had to change just a few characters to accommodate this extra level above `Policy`.  It might also reduce a lot of the small conflicts we have as arguments/fields are added here or there to these functions and structures.

---

How I got here: it seems like `PlanningInput` decomposes into:

- "policy" -- configuration and choices made by the operator
- "system state" -- miscellaneous stuff we have to fetch from the database for planning purposes

But right now `Policy` has `sleds` which contains sled policy (fine) but also sled state, zpools, and subnet, which by the definitions above are definitely "system state" and not "policy".  It seems like this second group should move into `PlanningInput`.  But then you have two maps of sleds, one for policy stuff and one for state.  Callers have to deal with one being present while the other's absent, etc.  If we put this behind an abstraction that ensures that these are consistent (e.g., constructor/builder + iterator that combines the information or whatever) then callers don't have to care how it's organized.

I also think we should add the internal and external DNS versions to `PlanningInput`.  Right now these are separate arguments to `BlueprintBuilder::build_initial_from_collection()`, `BlueprintBuilder::new_based_on()`, and `Planner::new_based_on()` but I don' see how they're any different from this other "system state" and it would clean up a lot of call sites to consolidate them into this struct.

There are also so many tests that make up bogus `Generation::new()` for the DNS versions because they don't care.  I didn't want to make these optional originally because I wanted people writing production code to be forced to think about what the generation should be.  But in practice there's basically one production caller and a ton of callers that variously care or don't care about this and it might be easier if it were just optional.

---

What do folks think?

Reply:
I'm with you for the vast majority of that, and would be happy to take a pass at it as a followup to this PR (and maybe poke at your suggestion about the relationship between `PlanningInput` and `UnstableReconfiguratorState` while I'm at it). One bit I'm not sure about:

> I also think we should add the internal and external DNS versions to PlanningInput

These are inputs to `BlueprintBuilder::build_initial_from_collection()`, but `PlanningInput` is not (at least on this PR). The extra system state stashed in `PlanningInput` is not relevant to build a blueprint from a collection, and especially as that state grows it seems reasonable (I think?) to avoid having to supply it unnecessarily.